lib/os/onoff.c

```diff
@@ -494,14 +512,22 @@ int onoff_cancel(struct onoff_service *srv,
 	/* Can't remove the last client waiting for the in-progress
 	 * transition, as there would be nobody to receive the
 	 * completion notification, which might indicate a service
-	 * error.
+	 * error. Unless there is a cancel function specified which
+	 * is capable to abort ongoing transition to on.
 	 */
 	if (sys_slist_find_and_remove(&srv->clients, &cli->node)) {
 		rv = 0;
 		if (sys_slist_is_empty(&srv->clients)
 		    && (state != SERVICE_STATE_TO_OFF)) {
-			rv = -EWOULDBLOCK;
-			sys_slist_append(&srv->clients, &cli->node);
+			if (srv->cancel && (state == SERVICE_STATE_TO_ON)) {
+				set_service_state(srv, SERVICE_STATE_TO_OFF);
```

Comment:
At this point, client is the last releaser, list is empty to accept new requests. transition to off is initiated. I'm not sure if `(state == SERVICE_STATE_TO_ON)`? If client is on the list, does it always mean that we are in volatile state (on to off or off to on)?